### PR TITLE
Keck HIRES continues

### PR DESCRIPTION
**Pull request by profxj**
This is another substantial PR on the path to fully enabling
Keck/HIRES reductions.

The code is now running end-to-end on the one test dataset
in the DevSuite.

I am submitting now to:

- Get eyes on the modest changes to the `WaveCalib` datamodel
- Have the refactoring of `ech_objfind()` ingested

This PR should not merge until after PR #1527

### CODE REVIEW

**Comment by profxj** [comment]
DevSuite tests are underway

File: pypeit/find_objects.py
Comment (jhennawi) [inline pypeit/find_objects.py]: Echelle object finding is a core functionality just like non-echelle object finding. The only object dependency is on specobjs. By adding the control flow that was in core here, one now needs all objects needed to instantiate find_objects, which adds more dependencies. Also, someone may just want to find objects but not sky-subtract, which this class does not allow one to do in a simple way. I don't see a current motivation for taking this echelle object finding control flow outside of core.

File: pypeit/wavecalib.py
Comment (jhennawi) [inline pypeit/wavecalib.py]: Can you add a few more words indicating what the detector image is. I.e. is it an image with an integer indicating the detector that contributed the pixel, etc.

**Review by jhennawi — CHANGES_REQUESTED** [review]
Nice work on this @profxj. The results look great. The only significant comment is the refactoring of the echelle object finding. I think that control flow should remain a function in core and not be moved into the find_objects class. 

Also, since the code was completely refactored, I cannot see what changes if any were actually made to the echelle object finding functionality via a normal side-by-side code review.  It may be useful to chat about what was changed if anything significant was changed.

File: pypeit/wavecalib.py
Comment (profxj) [reply pypeit/wavecalib.py]: done

File: pypeit/core/wavecal/echelle.py
Comment (kbwestfall) [inline pypeit/core/wavecal/echelle.py:190]: Give a short description.

File: pypeit/edgetrace.py
Comment (kbwestfall) [inline pypeit/edgetrace.py:985]: ?

File: pypeit/images/buildimage.py
Comment (kbwestfall) [inline pypeit/images/buildimage.py:30]: Do we need to save this image?  Or can we only require that we're able to reproduce it using the saved "DETECTOR" object?  If it's the image I'm thinking of, this should be fast to construct.

File: pypeit/par/pypeitpar.py
Comment (kbwestfall) [inline pypeit/par/pypeitpar.py]: I know you only moved this code, but I'm taking the opportunity to say that I don't understand why this is a parameter and not an attribute of the relevant spectrograph.  I.e., why do we need to toggle whether or not a spectrograph is a fixed-format echelle?

_Edit_: Actually, it looks like we do define this at least for the `KeckHIRESSpectrgraph`.  Can we consolidate?

File: pypeit/slittrace.py
Comment (kbwestfall) [inline pypeit/slittrace.py]: I'm not sure which is faster, but it seems like you could also use `numpy.unique`, setting `return_counts=True`, and return the unique value with the most counts.  The code would then be:
```
det, cnt = np.unique(det_img[(slit_img == spat_id) & (det_img > 0)], return_counts=True)
return det[np.argmax(cnt)]
```

File: pypeit/find_objects.py
Comment (kbwestfall) [inline pypeit/find_objects.py]: Please add a big bright flashing boldface underlined TODO that this needs to be settled before we merge to `release`.

Comment (jhennawi) [inline pypeit/find_objects.py]: Are you planning to delete the orig=False control flow below?

File: pypeit/spectrographs/keck_hires.py
Comment (jhennawi) [inline pypeit/spectrographs/keck_hires.py:62]: Should this be True now?

File: pypeit/spectrographs/keck_hires.py
Comment (jhennawi) [inline pypeit/spectrographs/keck_hires.py]: is this correct, i.e. gain is exactly 1.0? Also, I believe this depends on readout mode so we probably should put a TODO in there about that.

File: pypeit/spectrographs/keck_hires.py
Comment (jhennawi) [inline pypeit/spectrographs/keck_hires.py:546]: Now I'm confused. Rather than instantiate to bogus values and then set them again below, should we instantiate to None?

File: pypeit/wavecalib.py
Comment (jhennawi) [inline pypeit/wavecalib.py:581]: Can you add a TODO somewhere about merging this eventually with "reidentify" for fixed format echelles?

**Review by jhennawi — APPROVED** [review]
Hi @profxj, 

Thanks for the work here. I have a minor comment about the gain. The more significant comment is regarding the ech_objfind core routine. I like the changes you made there, and approve provided you don't revert back to the orig=False where that control flow is in the find_objects object.

**Comment by profxj** [comment]
--- PYTEST PYPEIT UNIT TESTS PASSED  206 passed, 137 warnings in 191.42s (0:03:11) ---
--- PYTEST UNIT TESTS PASSED  118 passed, 242 warnings in 885.83s (0:14:45) ---
--- PYTEST VET TESTS PASSED  30 passed, 10 warnings in 762.34s (0:12:42) ---
--- PYPEIT DEVELOPMENT SUITE FAILED 1/167 TESTS  ---
Failed tests:
    keck_hires/RED_C1_ECH_-0.82_XD_1.62 pypeit
Skipped tests:
Testing Started at 2023-02-17T05:15:50.969380
Testing Completed at 2023-02-17T13:50:11.598081
Total Time: 8:34:20.628701

Failed the main test! :)

File: pypeit/wavecalib.py
Comment (kbwestfall) [inline pypeit/wavecalib.py]: Can you articulate how in a comment so that we can look out for failure modes?

File: pypeit/wavecalib.py
Comment (kbwestfall) [inline pypeit/wavecalib.py]: If you don't want to remove it now, can you comment it out for now?

File: pypeit/wavecalib.py
Comment (kbwestfall) [inline pypeit/wavecalib.py]: (As annoying as it is) Can you use `logical_not` instead of `invert`?

File: pypeit/wavecalib.py
Comment (kbwestfall) [inline pypeit/wavecalib.py:677]: Is there a way to avoid having to recreate this so many times, but still not have to save it to the slit trace file?

File: pypeit/wavecalib.py
Comment (kbwestfall) [inline pypeit/wavecalib.py:692]: Will the elements of each of these empty arrays (i.e., the things you append below) always be the same shape?  If not, you might instead set `dtype=object`.  The allows the object to be an array of arrays.  I'm not sure how this is treated otherwise.

File: pypeit/core/findobj_skymask.py
Comment (kbwestfall) [inline pypeit/core/findobj_skymask.py:389]: Can `nfound` be 0?  If so, add an else here that throws an error?

File: pypeit/core/findobj_skymask.py
Comment (kbwestfall) [inline pypeit/core/findobj_skymask.py]: Does `_description_` do something?

File: pypeit/core/findobj_skymask.py
Comment (kbwestfall) [inline pypeit/core/findobj_skymask.py]: Skip the middle man: `return sobjs_trim.copy()`?

File: pypeit/core/findobj_skymask.py
Comment (kbwestfall) [inline pypeit/core/findobj_skymask.py]: Don't use `&` with single booleans.  Instead: 

`if spec.ech_frac_was_fit and spec.ech_snr > 1:`

File: pypeit/core/findobj_skymask.py
Comment (kbwestfall) [inline pypeit/core/findobj_skymask.py]: Instead of this `if..else` block, I would immediately return the `else` case so that the rest of it doesn't need to be within an if statement.  I.e.:
```
if len(sobjs_in_orders) == 0:
    return sobjs_in_orders

...

return ech_pca_traces(...)
```

**Review by kbwestfall — CHANGES_REQUESTED** [review]
Nice!  Nothing major from me.  I like the breaking up of the original function (at least what I was able to follow).  I put in a few comments that I realized afterward are likely hold-overs from the previous code.

File: pypeit/wavecalib.py
Comment (profxj) [reply pypeit/wavecalib.py]: done

File: pypeit/find_objects.py
Comment (profxj) [reply pypeit/find_objects.py]: yes

File: pypeit/spectrographs/keck_hires.py
Comment (profxj) [reply pypeit/spectrographs/keck_hires.py:62]: I don't think yet, e.g. we haven't implemented anything
sensible for flat fielding.  Am going to add notes here
on what else needs to be done before making it supported

File: pypeit/spectrographs/keck_hires.py
Comment (profxj) [reply pypeit/spectrographs/keck_hires.py]: set below

File: pypeit/spectrographs/keck_hires.py
Comment (profxj) [reply pypeit/spectrographs/keck_hires.py:546]: sure

File: pypeit/wavecalib.py
Comment (profxj) [reply pypeit/wavecalib.py:581]: added

File: pypeit/core/wavecal/echelle.py
Comment (profxj) [reply pypeit/core/wavecal/echelle.py:190]: added

File: pypeit/edgetrace.py
Comment (profxj) [reply pypeit/edgetrace.py:985]: linting was otherwise going bonkers.  no idea why..

File: pypeit/images/buildimage.py
Comment (profxj) [reply pypeit/images/buildimage.py:30]: I considered that, but to enable `WaveCalib`
to generate the wavelength image it needs
this and it may not have the rest of what is needed
to regenerate.

File: pypeit/par/pypeitpar.py
Comment (profxj) [reply pypeit/par/pypeitpar.py]: indeed, that `par` was superfluous
now removed

File: pypeit/slittrace.py
Comment (profxj) [reply pypeit/slittrace.py]: love it!  replaced

File: pypeit/find_objects.py
Comment (profxj) [reply pypeit/find_objects.py]: removing `orig` now

File: pypeit/wavecalib.py
Comment (profxj) [reply pypeit/wavecalib.py]: added

File: pypeit/wavecalib.py
Comment (profxj) [reply pypeit/wavecalib.py]: removed

File: pypeit/wavecalib.py
Comment (profxj) [reply pypeit/wavecalib.py]: done

File: pypeit/wavecalib.py
Comment (profxj) [reply pypeit/wavecalib.py:677]: I don't think so.  And because it needs to allow for flexure, etc.
this is a better approach.

File: pypeit/wavecalib.py
Comment (profxj) [reply pypeit/wavecalib.py:692]: good point, although as it ain't broke I'm not going to fix it.  ;)

File: pypeit/core/findobj_skymask.py
Comment (profxj) [reply pypeit/core/findobj_skymask.py:389]: Good call

File: pypeit/core/findobj_skymask.py
Comment (profxj) [reply pypeit/core/findobj_skymask.py]: removed

File: pypeit/core/findobj_skymask.py
Comment (profxj) [reply pypeit/core/findobj_skymask.py]: agreed

Comment (profxj) [reply pypeit/core/findobj_skymask.py]: and no `.copy()` necessary

File: pypeit/core/findobj_skymask.py
Comment (profxj) [reply pypeit/core/findobj_skymask.py]: modified

File: pypeit/core/findobj_skymask.py
Comment (profxj) [reply pypeit/core/findobj_skymask.py]: good call

**Comment by profxj** [comment]
Restarting tests

**Comment by profxj** [comment]
Passed all the tests except HIRES!  :) 

(pypeit) wolverine> tail xavier-keck-hires-3.report 
--- PYTEST PYPEIT UNIT TESTS PASSED  206 passed, 137 warnings in 211.40s (0:03:31) ---
--- PYTEST UNIT TESTS PASSED  118 passed, 242 warnings in 985.68s (0:16:25) ---
--- PYTEST VET TESTS PASSED  30 passed, 9 warnings in 865.38s (0:14:25) ---
--- PYPEIT DEVELOPMENT SUITE FAILED 1/167 TESTS  ---
Failed tests:
    keck_hires/RED_C1_ECH_-0.82_XD_1.62 pypeit
Skipped tests:
Testing Started at 2023-02-19T07:53:14.048351
Testing Completed at 2023-02-19T16:35:53.630866
Total Time: 8:42:39.582515

I introduced a bug updating `gain` handling as per the PR recommendation.
Fixed now

**Comment by profxj** [comment]
Passed:

(pypeit) wolverine> ./pypeit_test -i keck_hires -t 4 reduce
Running reduce tests.
Running tests on the following instruments:
    keck_hires

Reducing data from keck_hires for the following setups:
    RED_C1_ECH_-0.82_XD_1.62

Running tests in 4 parallel processes
 0 passed/ 0 failed/ 0 skipped STARTED keck_hires/RED_C1_ECH_-0.82_XD_1.62 pypeit
 1 passed/ 0 failed/ 0 skipped PASSED  keck_hires/RED_C1_ECH_-0.82_XD_1.62 pypeit
--- PYPEIT DEVELOPMENT SUITE PASSED 1/1 TESTS  ---

Test Summary
--------------------------------------------------------
--- PYPEIT DEVELOPMENT SUITE PASSED 1/1 TESTS  ---
Testing Started at 2023-02-19T13:18:51.559156
Testing Completed at 2023-02-19T14:50:15.635608
Total Time: 1:31:24.076452

**Review by kbwestfall — APPROVED** [review]

**Pull request merged**
